### PR TITLE
now pretty print works for scaler variables too.

### PR DESCRIPTION
**Pull request by mohammadpz**
Hi @dwf ,
Here I just handle output[0] of a FunctionGraph in a Composite.
I'm now sure how should I print multiple outputs.
I'm still struggling with a good test file. The only thing I found is this test:

import numpy
import theano

def detect_nan(i, node, fn):
    for output in fn.outputs:
        if numpy.isnan(output[0]).any():
            print '*** NaN detected ***'
            theano.printing.debugprint(node)
            print 'Inputs : %s' % [input[0] for input in fn.inputs]
            print 'Outputs: %s' % [output[0] for output in fn.outputs]
            break

x = theano.tensor.dscalar('x')
f = theano.function([x], [theano.tensor.log(x) * x],
                    mode=theano.compile.MonitorMode(
                        post_func=detect_nan))
f(0)

### CODE REVIEW

**Comment by lamblin** [comment]
This is related to issue gh-2134. @mohammadpz, in the future, can you refer to the ticket number in a comment? Thanks.

**Comment by mohammadpz** [comment]
@lamblin ,
Sorry. I didn't know that.

**Comment by dwf** [comment]
None of the indentation is preserved in the code you pasted in the comment. Put it in a block with \`\`\`python on a line at the beginning and \`\`\` on a line at the end (see the [Markdown instructions](https://guides.github.com/features/mastering-markdown/), also linked above the comment box). Right now I can't easily tell what the code does because the scopes have all been collapsed.

**Comment by dwf** [comment]
Your test should probably just construct some semi-complicated expressions that make use of this functionality and then call `str()`, and verify the string is what it ought to be.  Don't overthink it.

**Comment by mohammadpz** [comment]
Here it is:
```import numpy

import theano

def detect_nan(i, node, fn):
    for output in fn.outputs:
        if numpy.isnan(output[0]).any():
            print '*** NaN detected ***'
            theano.printing.debugprint(node)
            print 'Inputs : %s' % [input[0] for input in fn.inputs]
            print 'Outputs: %s' % [output[0] for output in fn.outputs]
            break

x = theano.tensor.dscalar('x')
f = theano.function([x], [theano.tensor.log(x) * x],
                    mode=theano.compile.MonitorMode(
                        post_func=detect_nan))
f(0)```

**Comment by dwf** [comment]
Yeah, that code isn't really relevant. I don't think k you should need to fiddle with graph node internals at all. What you should do is

* construct a Composite instance (with an appropriate inner graph that operates on scalars)
* verify that what you get when you call *str()* on that thing looks right
* copy that code, and that output into a test case (see any file in Theano that starts with test_) and have an assertion the that the expected output is what you get
* repeat for several kinds of inner graphs and all arithmetic operations

And you're done

**Comment by dwf** [comment]
If it's the precise way to write a test that you're having trouble with, have a look at this [nose introduction](http://pythontesting.net/framework/nose/nose-introduction/).

**Comment by mohammadpz** [comment]
Using this code I made different Composite instance:
```python
import numpy
import theano

x = theano.tensor.dscalar('x')

f = theano.function([x], [theano.tensor.log(x) * x])
theano.printing.debugprint(f)

f = theano.function([x], [theano.tensor.log(x) + x])
theano.printing.debugprint(f)

f = theano.function([x], [theano.tensor.log(x) - x])
theano.printing.debugprint(f)

f = theano.function([x], [theano.tensor.log(x) / x])
theano.printing.debugprint(f)

f = theano.function([x], [theano.tensor.log(x) // x])
theano.printing.debugprint(f)

f = theano.function([x], [theano.tensor.log(x) ** x])
theano.printing.debugprint(f)

f = theano.function([x], [-theano.tensor.log(x)])
theano.printing.debugprint(f)
```
Before this pull request the result looks like:
```
-> python t.py
Elemwise{Composite{[mul(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[add(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[sub(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[true_div(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[int_div(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[pow(log(i0), i0)]}} [@A] ''   0
 |x [@B]
Elemwise{Composite{[neg(log(i0))]}} [@A] ''   0
 |x [@B]
```
After this pull request, result look like:
```
-> python t.py
Elemwise{Composite{(log(i0) * i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(log(i0) + i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(log(i0) - i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(log(i0) / i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(log(i0) // i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(log(i0) ** i0)}} [@A] ''   0
 |x [@B]
Elemwise{Composite{(-log(i0))}} [@A] ''   0
 |x [@B]
```

The page about Composite in deeplearning.net is empty. I don't know how should I make a Composite with multiple outputs in its function graph.

**Comment by nouiz** [comment]
There is an example in the file scalar/tests/test_basic.py, grep for the
method test_many_outputs

On Sun, Jan 18, 2015 at 7:12 PM, Mohammad Pezeshki <notifications@github.com

**Comment by dwf** [comment]
The Composite op generally isn't created by the user, which is why the
documentation is sparse, but as Fred said, there are examples in the tests.

More generally, reading the code in order to figure that stuff out is part
of your ticket assignment. If you're really stuck then by all means, say
so, but part of the point of the CCW is to get you familiar with reading
code you don't quite understand yet. This is a valuable skill, in academia
and in industry, that you can only really develop through practice.

**Comment by mohammadpz** [comment]
@nouiz , Thanks for your comment.
@dwf , I want you to know that I'm doing my best! It's just 3 months that I'm getting familiar with python.
In my previous CCW task, I got this feedback from @vdumoulin that I should always ask if anything is not clear for you.
By the way, thanks @dwf for your note. I try to perform better in my next tasks.

I updated my pull request and added a method ```test_composite_printing``` in ```scalar/tests/test_basic.py```.
Here's the output of ```str()``` on a Composite before pull request:
```
[*1 -> Composite{[add(add(i0, i1), i2), add(i0, mul(i1, i2)), true_div(i0, i1)]}(x, y, z), *1::1, *1::2]
```
And here's the output of ```str()``` after pull request:
```
[*1 -> Composite{((i0 + i1) + i2), (i0 + (i1 * i2)), (i0 / i1)}(x, y, z), *1::1, *1::2]
```

**Comment by dwf** [comment]
No problem. You *should* ask questions, if you don't understand something, but there's a difference between not understanding what the code is doing and just looking at the documentation and concluding an explanation isn't there. :)

If you can add a few more test cases that would be good.

I see a few ( //, ** ) that aren't tested. You probably need integer inputs for //, not sure. Also what about % (mod)? It doesn't seem to be mentioned...

Once that's done and there are a few more tests of different expressions, I think we can call it finished.

File: theano/scalar/tests/test_basic.py [hunk 1]
```diff
@@ -146,6 +146,17 @@ def test_many_outputs(self):
...
+    def test_composite_printing(self):
+        x, y, z = floats('xyz')
+        e0 = x + y + z
+        e1 = x + y * z
+        e2 = x / y
+        C = Composite([x, y, z], [e0, e1, e2])
+        c = C.make_node(x, y, z)
+        g = FunctionGraph([x, y, z], c.outputs)
+        fn = gof.DualLinker().accept(g).make_function()
+        assert str(g) == '[*1 -> Composite{((i0 + i1) + i2), (i0 + (i1 * i2)), (i0 / i1)}(x, y, z), *1::1, *1::2]'
```
Comment (dwf) [inline theano/scalar/tests/test_basic.py]: Also, put parentheses around  the string and break it in the middle, like

```python
('This is too long to fit on one line, and so we '
 'will instead break it across multiple lines.')
```
Make sure the quotation marks line up as per PEP8.

**Comment by mohammadpz** [comment]
@dwf , I really appreciate any kind of comment which improve my skills. Thanks :)

I added more tests.
I also added ```mod``` and it works. But I don't know why it isn't managed it ```tensor/basic.py```. Should I open an issue?

**Comment by dwf** [comment]
Might be managed under the name "divmod" or something. Anyway, this looks
good, thanks. Merging.

On Mon, Jan 19, 2015 at 5:58 PM, Mohammad Pezeshki <notifications@github.com

**Comment by mohammadpz** [comment]
Thank you for your supervision during this task.

**Comment by dwf** [comment]
This pull request is failing Travis but it seems that the failure is unrelated (timeout in the scan tests) so I'm merging anyway.

**Comment by mohammadpz** [comment]
Failure on travis happens time to time.

File: theano/scalar/basic.py [hunk 2]
```diff
@@ -2166,6 +2168,14 @@ def c_code(self, node, name, (x,), (z,), sub):
...
 neg = Neg(same_out, name='neg')
 
+pprint.assign(add, printing.OperatorPrinter('+', -2, 'either'))
+pprint.assign(mul, printing.OperatorPrinter('*', -1, 'either'))
+pprint.assign(sub, printing.OperatorPrinter('-', -2, 'left'))
+pprint.assign(neg, printing.OperatorPrinter('-', 0, 'either'))
+pprint.assign(true_div, printing.OperatorPrinter('/', -1, 'left'))
+pprint.assign(int_div, printing.OperatorPrinter('//', -1, 'left'))
+pprint.assign(pow, printing.OperatorPrinter('**', 1, 'right'))
+pprint.assign(mod, printing.OperatorPrinter('%', -1, 'lef'))
```
Comment (lamblin) [inline theano/scalar/basic.py]: Should be `'left'`, not `'lef'`, I think.

File: theano/scalar/basic.py [hunk 3]
```diff
@@ -3099,7 +3109,8 @@ def init_name(self):
                 for i, r in enumerate(self.fgraph.variables):
                     if r not in io and len(r.clients) > 1:
                         r.name = 't%i' % i
-                rval = "Composite{%s}" % str(self.fgraph)
+                rval = "Composite{%s}" % ', '.join([pprint(output) for output
+                                                    in self.fgraph.outputs])
```
Comment (nouiz) [inline theano/scalar/basic.py]: I'm fine with this, but if there is many outputs, this hide the reuse of the different part. Just a comment so that if you see this, we know it is just they way it is printed.

I looked at the code of pprint() and it would accept this: pprint(self.fgraph.outputs). This would remove this "problem".

Comment (nouiz) [inline theano/scalar/basic.py]: I didn't look at the output, maybe it make it ugly, so I let you try it and decide what to do. If you do not do it, just put in a comment the output that you find ugly.

File: theano/scalar/basic.py (hunk 2 above)
Comment (dwf) [inline theano/scalar/basic.py]: Oops, I missed that one. @mohammadpz Can you send another PR to fix the
typo?

Comment (mohammadpz) [inline theano/scalar/basic.py]: @lamblin , @dwf , Sorry for typo. I'll fix it.

File: theano/scalar/basic.py (hunk 3 above)
Comment (mohammadpz) [inline theano/scalar/basic.py]: @nouiz , I'm not sure that I understand your comment.
But, when I write ```pprint(self.fgraph.outputs)``` I get this error:
```python
AttributeError: 'list' object has no attribute 'owner'
```
By "ugly output", what exactly you mean?

File: theano/scalar/basic.py (hunk 2 above)
Comment (dwf) [inline theano/scalar/basic.py]: This raises the issue that OperatorPrinter doesn't check for valid kinds of
associativity.

@mohammadpz While you're fixing the typo can you add an assertion here
https://github.com/Theano/Theano/blob/master/theano/printing.py#L277 that
checks assoc is either "left", "right" or "either"? Good idea to define
VALID_ASSOC = set(['left', 'right', 'either']) at the class level and then
check that "assoc in VALID_ASSOC". This assertion would've caught the typo.

On Wed, Jan 21, 2015 at 2:51 PM, Mohammad Pezeshki <notifications@github.com

File: theano/scalar/basic.py (hunk 3 above)
Comment (nouiz) [inline theano/scalar/basic.py]: By ugly output, I mean that we see the same computation multiple time. But
as my solution do not work, we can keep it like this. Just remember that
fact in case people have question about this.

On Wed, Jan 21, 2015 at 3:12 PM, Mohammad Pezeshki <notifications@github.com
